### ISSUE TITLE
coord.Distance with pd.options.display.max_columns set produces UnitConversionError on Jupyter Notebook

### ISSUE THREAD

**Issue by jan-niedospial** [astropy/astropy#12670]
### Description

coord.Distance produces UnitConversionError when pd.options.display.max_columns is set, for example to 100. 

When converting distance modulus to distance and trying to see the dataframe.head() with custom max_columns settings, it will produce UnitConversionError on Jupyter Notebook.

### Expected behavior

I would expect to be able to see more columns in a dataframe in a preview.

### Actual behavior

UnitConversionError. It is very confusing, because it's not there with default columns display settings.

### Steps to Reproduce

1. Import libraries
2. Set pandas options to display 300 columns
3. Load data with distance modulus (for example, from Hyperleda meandata with modbest - http://leda.univ-lyon1.fr/fullsql.html - 10 records sample attached), as df 
4. Convert the distance modulus to distance
5. Try to see top 5 rows of the dataframe

Repeat the same with default display settings to check if it works (it should).

```python
# Put your Python code snippet here.

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
from scipy import integrate

import math as m
from math import sqrt

import astropy.coordinates as coord
import astropy.units as u
from astropy.io import ascii
from astropy.coordinates import SkyCoord

# columns display settings
pd.options.display.max_columns=100

# loading df
df = pd.read_csv("data_10.csv")

# converting distance modulus to distance in parsecs
df['dist'] = coord.Distance(distmod=df['modbest'], unit='pc')

# checking the dataframe
df.head()

```

### System Details

Windows-10-10.0.19041-SP0
Python 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
Numpy 1.21.2
pyerfa 2.0.0
astropy 5.0
Scipy 1.7.1
Matplotlib 3.4.3
Pandas 1.3.4

I also checked on older versions of Python, Numpy, Pandas (1.3.3), astropy (4.02, 4.2) and it was the same
[data_10.csv](https://github.com/astropy/astropy/files/7791195/data_10.csv)
.

**Comment by pllim** [comment]
Have you raised this issue in `pandas`? We never promised flawless integration between `pandas.DataFrame` and `astropy.units`. I am not sure where the fix should be, or if this is even a supported use case. `astropy.units` will/should integrate properly with `astropy.table.QTable` though.

**Comment by mhvk** [comment]
@jan-niedospial - it is helpful to provide a more minimal example. Here, it seems clear that display is the problem:
```
import astropy.coordinates as coord
import pandas as pd
df = pd.DataFrame()
df['dist'] = coord.Distance([6e8], 'pc')
df
.
.
.
~/data/mhvk/packages/astropy/.tox/test/lib/python3.9/site-packages/pandas/io/formats/format.py in get_result_as_array(self)
   1500             # large values: more that 8 characters including decimal symbol
   1501             # and first digit, hence > 1e6
-> 1502             has_large_values = (abs_vals > 1e6).any()
.
.
.
UnitConversionError: Can only apply 'greater' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)
```
With that, one sees that things go wrong for any quantity (`df['dist'] = u.Quantity([10], 'm')` fails too), because in its printing pandas *assumes* that a `Series` contains pure numbers which can be compared to other pure numbers, while `Quantity` underneath tries to compare something in parsec to something dimensionless. Looking at the pandas code, it seems that one could prevent this from happening by not using fixed format for display, but I do not know enough about pandas to try that.

Anyway, bottom line is that sadly quantities cannot be used very well with `pandas`. You could either add `.value` to your `Distance` line and just keep track yourself that the unit is 'pc', or, as @pllim suggested, try `astropy.table.QTable` instead.

**Comment by jan-niedospial** [comment]
@mhvk Thank you for that. I will try to provide more minimal example if there is next time :)

@pllim I didn't log it for pandas, I didn't know where the error lies either. 

The issue is not severe, at least not for me - there are simple workarounds ( @mhvk ".value" is a very useful tip indeed, thank you! - much faster than casting as string and removing the unit, then converting back to float) or just formulas that can be used instead. 

The problem I see is that this incompatibility between pandas and astropy is not stated anywhere directly. As both astropy and pandas are in the standard library, I worked under the (incorrect) assumption that they should work well together. From a new user perspective, coming from pandas, trying to figure out what is going on can be mind bending. I couldn't find anything about that on Github (probably searched for the wrong thing), so I logged it as a bug to let the devs know there is an issue, in case you didn't know.

@pllim thanks to you I've just learned about the existence of  astropy.table.QTable, I will try it out.

Thank you both, it's very helpful.

**Comment by pllim** [comment]
Yes, the incompatibility is unfortunate, but `astropy.table.Table` (and `QTable` too maybe) tries to be nice with providing `from_pandas` and `to_pandas` method. If `pandas` devs want to work with us to support `astropy.units`, I think that would be very welcome!

**Comment by mhvk** [comment]
@jan-niedospial - no worries, these issues are good even if only so there is something to find for others.... In principle, for cases such as this perhaps more documentation would help, but am not sure where to put it -- there is already so much that the chance has become small that someone encounters it...

**Comment by jan-niedospial** [comment]
@pllim I've read through some of the many threads in pandas Github about units support, and it seems they are unwilling to implement units support for many reasons, at least not before pandas 2. I've seen you rasing this issue before to no avail.

@mhvk I would agree, with extensive docs it would be hard to find. If I was to suggest something, it would be a warning.

I've tried to see if it's possible to implement, and I came up with something like this (and I know it's way too primitive):

```
def printer(x):

    try:
        if isinstance(x, pd.DataFrame):
            import warnings
            warnings.warn("You are trying to apply operation with units in pandas DataFrame, which doesn't support units." 
            "Consider using '.value' method to remove units, or 'use astropy.table.QTable' instead.")
    except:
        pass
    
    print(x)
```
This route, from pandas to astropy, may be unusual, but for those that happen to follow it, this warning would save some time and nerves.

Alas, I can also see how implementing it may simply not be worth it, especially when you consider that in the future verion of pandas which actually supports units, it will probably be still "pd.DataFrame".  It would also not work 100% of the time, if people import pandas as anything else as pd. 

I will just leave it as food for thought ;)

Happy New Year.

**Comment by mhvk** [comment]
@jan-niedospial - thanks for the suggestion. The problem is that from within astropy we have no control over any of this - it all gets done inside pandas. 

I guess one option that perhaps might make sense is to have a general documentation page about interaction with other big packages such as `pandas`, `scipy`, `matplotlib`, `dask`, `xarray`, etc., that perhaps collects tips and tricks (e.g., the visualization stuff for `matplotlib`).

**Comment by jan-niedospial** [comment]
@mhvk I see. It makes sense.

That sounds good. Thank you very much for all of this.
